### DISCUSSION TITLE
Anonymous recursion

### DISCUSSION THREAD

**Discussion by raviqqe** [pen-lang/pen#786] [Ideas]
# Problem

- Currently, every recursion (iteration, loops, etc.) needs to be named because it needs to be a top-level function.

# Solution

- Add anonymous recursion syntax.
  - There are some examples found in other languages like Clojure and Scala.
- Downsides are:
  - We have two ways to do one thing.
  - It's not really saving anything in terms of lines of codes.
- Should we limit this construct's use cases to strict evaluation and keep list comprehension?

### Map

- This is a kind of useless example as we can simply unwrap the `for` loop.

```pen
squareMany = \(xs [number]) [number] {
  for {
    if [x, ...xs] = xs {
      [number x() * x(), ...recur]
    } else {
      [number]
    }
  }
}
```

### Reduce

```pen
sum = \(xs [number]) number {
  y = 0

  for {
    if [x, ...xs] = xs {
      y = y + x()
      recur
    } else {
      y
    }
  }
}
```

### Filter

```pen
squareMany = \(xs [number]) [number] {
  for {
    if [x, ...xs] = xs {
      if x() > 42 {
        [number x(), ...recur]
      } else {
        recur
      }
    } else {
      [number]
    }
  }
}
```

### Zip

```pen
zip = \(xs [number], ys [number]) [number] {
  for {
    if [x, ...xs] = xs {
      if [y, ...ys] = ys {
        [number x() + y(), ...recur]
      } else {
        [number]
      }
    } else {
      [number]
    }
  }
}
```

# References

- [`loop` | ClojureDocs](https://clojuredocs.org/clojure.core/loop)
- [`for` loops | Scala](https://docs.scala-lang.org/overviews/scala-book/for-loops.html)
- #1353

**Comment by sammy-hughes** [comment]
Is the intent here that the standard function "recur" would refer to the current scope?

**Reply by raviqqe** [reply]
Yes! I just added a References section that contains other languages' examples.

**Comment by sammy-hughes** [comment]
I was set to request a reducer combinator, but that approach absolutely satisfies any use-case I can think of!
